data/blog/libtailscale.mdx

```diff
@@ -0,0 +1,198 @@
+Those rules are designed to allow the
+[precise collector](https://en.wikipedia.org/wiki/Tracing_garbage_collection#Precise_vs._conservative_and_internal_pointers)
+to know where all pointers to Go memory are at all times, and to allow
+the Go runtime implementation to switch to a moving collector in the
+future if the trade-offs ever make sense for the sorts of programs
+written in Go.
+The result is a set of tight restrictions on when C is allowed to hold
+a pointer into Go memory.
+There is also some degree of enforcement of these rules in Go runtime
+checks, to help people from writing programs that might break
+surprisingly in the future with changes in Go.
+
+The full details of Go’s rules for passing pointers are in the
+[cgo documentation](https://pkg.go.dev/cmd/cgo#hdr-Passing_pointers).
+Two things stand out:
+
+- _Go code may pass a Go pointer to C provided the Go memory to which it points does not contain any Go pointers._
+
+- _C code may not keep a copy of a Go pointer after the call returns._
+
+These are very limiting!
+(Again, for good reasons, both today and in the future.)
+
+The first rule, no pointers to pointers, means that only simple memory
+can be passed from Go to C.
+For example, you could pass a `*byte` pointer into a slice and have C
+read out of the slice, or write into it.
+That can be very useful for efficiently copying large amounts of data.
+But if you have an `*http.Server` you can never pass a reference to C,
+for it to then use exported functions to call methods on the object.
+
+Combined with the second rule, it becomes effectively impossible to
+implement something that looks like a socket API directly as we do
+inside the Go `tsnet` package:
+
+```
+tailscale tailscale_new();
+int tailscale_dial(tailscale sd, const char* network, const char* addr, tailscale_conn* conn_out);
+```
+
+So how do we do it?
+
+## Handles
+
+It turns out we have a really well-trod conceptual model for
+interacting from C with a managed memory space: system calls.
+Kernel memory is generally off-limits to userspace and needs to be
+operated on through intermediary functions.
+Userspace memory is accessible to the kernel, but because kernels do
+not have a lot of insight into how userspace manages its memory,
+passing a pointer is always done under very carefully specified conditions.
+
+The typical way to refer to a kernel object from userspace is to be
+given a handle.
+The most familiar handle to a Unix user is a
+[file descriptor](https://en.wikipedia.org/wiki/File_descriptor).
+These are process-unique integers managed that refer to an entry table
+of objects stored in the kernel.
+
+We can do exactly the same thing for Go objects:
+
+```
+// servers tracks all the allocated *tsnet.Server objects.
+var servers struct {
+        mu   sync.Mutex
+        next C.int
+        m    map[C.int]*tsnet.Server
+}
+```
+
+An Go function `//export`ed to C can then allocate a new `*tsnet.Server`,
```

Comment:
```suggestion
A Go function `//export`-ed to C can then allocate a new `*tsnet.Server`,
```

Reply:
Check to see whether the rendered `//export`-ed phrase introduces an extra space (between the code portion and the plain text portion).

Reply:
`//export` is odd but correct. It's specific to cgo and C references to Go.

Reply:
![image](https://user-images.githubusercontent.com/79330/222307911-f28155bd-f58b-4f5a-a004-e20b4088a1a2.png)

Reply:
It's passable 🤷